numpyro/distributions/continuous.py

```diff
@@ -651,6 +651,151 @@ def variance(self):
+    Both `bmat` and `bvec` may have any number of leading dimensions, which correspond
+    to a batch shape. They are not necessarily assumed to have the same batch shape,
+    just ones which can be broadcasted.
+    """
+    return np.squeeze(np.matmul(bmat, np.expand_dims(bvec, axis=-1)), axis=-1)
+
+
+def _batch_capacitance_tril(W, D):
+    r"""
+    Computes Cholesky of :math:`I + W.T @ inv(D) @ W` for a batch of matrices :math:`W`
+    and a batch of vectors :math:`D`.
+    """
+    Wt_Dinv = np.swapaxes(W, -1, -2) / np.expand_dims(D, -2)
+    K = np.matmul(Wt_Dinv, W)
+    # could be inefficient
+    return np.linalg.cholesky(np.add(K, np.identity(K.shape(-1))))
+
+
+def _batch_lowrank_logdet(W, D, capacitance_tril):
+    r"""
+    Uses "matrix determinant lemma"::
+        log|W @ W.T + D| = log|C| + log|D|,
+    where :math:`C` is the capacitance matrix :math:`I + W.T @ inv(D) @ W`, to compute
+    the log determinant.
+    """
+    return 2 * np.sum(np.log(np.diagonal(capacitance_tril, axis1=-2, axis2=-1)), axis=-1) + np.log(D).sum(-1)
+
+
+def _batch_lowrank_mahalanobis(W, D, x, capacitance_tril):
+    r"""
+    Uses "Woodbury matrix identity"::
+        inv(W @ W.T + D) = inv(D) - inv(D) @ W @ inv(C) @ W.T @ inv(D),
+    where :math:`C` is the capacitance matrix :math:`I + W.T @ inv(D) @ W`, to compute the squared
+    Mahalanobis distance :math:`x.T @ inv(W @ W.T + D) @ x`.
+    """
+    Wt_Dinv = np.swapaxes(W, -1, -2) / np.expand_dims(D, -2)
+    Wt_Dinv_x = _batch_mv(Wt_Dinv, x)
+    mahalanobis_term1 = np.sum(np.square(x) / D, axis=-1)
+    mahalanobis_term2 = _batch_mahalanobis(capacitance_tril, Wt_Dinv_x)
+    return mahalanobis_term1 - mahalanobis_term2
+
+
+@copy_docs_from(Distribution)
+class LowRankMultivariateNormal(Distribution):
+    arg_constraints = {
+        "loc": constraints.real_vector,
+        "cov_factor": constraints.real,
+        "cov_diag": constraints.positive
+        }
+    support = constraints.real_vector
+
+    def __init__(self, loc, cov_factor, cov_diag, validate_args=None):
+        if np.ndim(loc) < 1:
+            raise ValueError("`loc` must be at least one-dimensional.")
+        event_shape = np.shape(loc)[-1:]
+        if np.ndim(cov_factor) < 2:
+            raise ValueError("`cov_factor` must be at least two-dimensional, "
+                             "with optional leading batch dimensions")
+        if np.shape(cov_factor)[-2:-1] != event_shape:
+            raise ValueError("`cov_factor` must be a batch of matrices with shape {} x m"
+                             .format(event_shape[0]))
+        if np.shape(cov_diag)[-1:] != event_shape:
+            raise ValueError("`cov_diag` must be a batch of vectors with shape {}".format(self.event_shape))
+
+        loc, cov_factor, cov_diag = promote_shapes(loc[..., np.newaxis], cov_factor, cov_diag[..., np.newaxis])
+        batch_shape = lax.broadcast_shapes(np.shape(loc), np.shape(cov_factor), np.shape(cov_diag))[:-2]
```

Comment:
sorry, my last code snippet missed a line after this
```
cov_diag = cov_diag[..., 0]
```

Reply:
Do you mean, `self.cov_diag = cov_diag[..., 0]` ? But i think you mentioned earlier that cov_diag is usually a vector (since it represents a diagonal matrix) -- so what's the intention behind slicing for the first column here ? (just out of curiosity)

Reply:
Ah, because we added a new axis `cov_diag[..., np.newaxis]` above (to promote_shapes), we drop it here. ^^

Reply:
@TuanNguyen27 I think the issue happens at this line. It is better to do `cov_diag = cov_diag[..., 0]` here, rather than in `self.cov_diag = cov_diag[..., 0]` because we use `cov_diag` to compute _batch_capacitance_tril.